tools/bindsnoop.py

```diff
@@ -0,0 +1,516 @@
+    help="trace cgroups in this BPF map only")
+parser.add_argument("--ebpf", action="store_true",
+    help=argparse.SUPPRESS)
+parser.add_argument("--debug-source", action="store_true",
+    help=argparse.SUPPRESS)
+args = parser.parse_args()
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wtautological-compare"
+#include <net/sock.h>
+#pragma clang diagnostic pop
+#include <net/inet_sock.h>
+#include <net/net_namespace.h>
+#include <bcc/proto.h>
+
+BPF_HASH(currsock, u32, struct socket *);
+
+// separate data structs for ipv4 and ipv6
+struct ipv4_bind_data_t {
+    u64 ts_us;
+    u32 pid;
+    u32 uid;
+    u64 ip;
+    u32 saddr;
+    u32 bound_dev_if;
+    int return_code;
+    u16 sport;
+    u8 socket_options;
+    u8 protocol;
+    char task[TASK_COMM_LEN];
+};
+BPF_PERF_OUTPUT(ipv4_bind_events);
+
+struct ipv6_bind_data_t {
+    // int128 would be aligned on 16 bytes boundary, better to go first
+    unsigned __int128 saddr;
+    u64 ts_us;
+    u32 pid;
+    u32 uid;
+    u64 ip;
+    u32 bound_dev_if;
+    int return_code;
+    u16 sport;
+    u8 socket_options;
+    u8 protocol;
+    char task[TASK_COMM_LEN];
+};
+BPF_PERF_OUTPUT(ipv6_bind_events);
+
+// separate flow keys per address family
+struct ipv4_flow_key_t {
+    u32 saddr;
+    u16 sport;
+};
+BPF_HASH(ipv4_count, struct ipv4_flow_key_t);
+
+struct ipv6_flow_key_t {
+    unsigned __int128 saddr;
+    u16 sport;
+};
+BPF_HASH(ipv6_count, struct ipv6_flow_key_t);
+
+CGROUP_MAP
+
+// bind options for event reporting
+union bind_options {
+    u8 data;
+    struct {
+        u8 freebind:1;
+        u8 transparent:1;
+        u8 bind_address_no_port:1;
+        u8 reuseaddress:1;
+        u8 reuseport:1;
+    } fields;
+};
+
+// TODO: add reporting for the original bind arguments
+int bindsnoop_entry(struct pt_regs *ctx, struct socket *socket)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = pid_tgid;
+    FILTER_PID
+
+    u32 uid = bpf_get_current_uid_gid();
+
+    FILTER_UID
+
+    FILTER_CGROUP
+
+    // stash the sock ptr for lookup on return
+    currsock.update(&tid, &socket);
+
+    return 0;
+};
+
+
+static int bindsnoop_return(struct pt_regs *ctx, short ipver)
+{
+    int ret = PT_REGS_RC(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = pid_tgid;
+
+    struct socket **skpp;
+    skpp = currsock.lookup(&tid);
+    if (skpp == 0) {
+        return 0;   // missed entry
+    }
+
+    int ignore_errors = 1;
+    FILTER_ERRORS
+    if (ret != 0 && ignore_errors) {
+        // failed to bind
+        currsock.delete(&tid);
+        return 0;
+    }
+
+    // pull in details
+    struct socket *skp_ = *skpp;
+    struct sock *skp = skp_->sk;
+
+    struct inet_sock *sockp = (struct inet_sock *)skp;
+
+    u16 sport;
+    bpf_probe_read(&sport, sizeof(sport), &sockp->inet_sport);
```

Comment:
Could you initialize sport say to `sport = 0`? ubuntu 16.04, which has kernel 4.4, will fail with verifier due to this.

Reply:
Added initialization. I'll need to create some testing environment to catch it in the future.